### DISCUSSION TITLE
Using safeParse has unknown output compared to parse which uses schema type

### DISCUSSION THREAD

**Discussion by thasmo** [open-circle/valibot#698] [Q&A] [answered]
Maybe I'm doing something wrong, but here `output` is typed with the schema I pass:

```ts
const output = v.parse(schema, data);
// output has infered `schema` type
```

But this `output` is of type `unknown`:
```ts
const { output } = v.safeParse(schema, data);
// output has `unknown` type
```

Is this intentional, or am I missing something here? I would assume `output` to be of type `schema` as well.

Thank you!

**Comment by fabian-hiller** [comment]
`safeParse` does not throw if there are any issues. Therefore, you have to check `.success` or `.issues` first. If there is no issue, `.output` will be typed similar to `parse`.

```ts
const result = v.safeParse(schema, data);

if (!result.issues) {
 // Your code here
}
```

**Reply by thasmo** [reply]
I'm trying this:

```ts
const schema = v.object({
	DEBUG: v.union([v.literal('true'), v.literal('false')]),
});

const { success, output } = v.safeParse(schema, process.env);

if (!success) {
	throw new Error(`Environment variables must be configured correctly.`);
}

export type EnvironmentSchema = v.InferOutput<typeof schema>;
export default output;
```

But I have to use `satisfies` to get the exported value to be typed properly:

```ts
const schema = v.object({
	DEBUG: v.union([v.literal('true'), v.literal('false')]),
});

const { success, output } = v.safeParse(schema, process.env);

if (!success) {
	throw new Error(`Environment variables must be configured correctly.`);
}

export type EnvironmentSchema = v.InferOutput<typeof schema>;
export default output satisfies EnvironmentSchema;
```

Maybe this is more of a TS issue/topic and does not belong to `valibot` directly. 🤔

**Reply by fabian-hiller** [reply, marked as answer]
This is because you have destructured the retuned object of `safeParse`. Try it without it, as shown in my example. TypeScript has no way of knowing that they are related if you destructure the object.

**Reply by thasmo** [reply]
Thanks a lot; this was it! Appreciate your time! 👍
